In [1]:
// подключаем необходимые библиотеки
import $ivy.`org.scalanlp::breeze:1.0`
import scala.collection.mutable.ArrayBuffer
import breeze.linalg._

import $ivy.$                         

import scala.collection.mutable.ArrayBuffer

import breeze.linalg._

In [2]:
// читаем файл с данными
val dataFile = io.Source.fromFile("Dataset.csv")

// подготавливаем матрицу
var dataMatrix = DenseMatrix.zeros[Double](0,28)

var i = 0
// Пробегаемся по каждай строке файла
for (line <- dataFile.getLines.drop(1)) { 
    
    // готовим массив для дальнейшего его перегона в матрицу 
    var dataArray = Array[Double]()
    
    // делим строку на элементы и обрезаем пробелы
    for (elem <- line.split(",")map(_.trim)){
        // если значения нет, то добавляем 0
        if(elem == "") {
            dataArray = dataArray :+ 0.0 
        // иначе добавляем в массив значение елемента
        }else{
            dataArray = dataArray :+ elem.toDouble
        }
    }
    
    // наполняем матрицу
    dataMatrix = DenseMatrix.vertcat(dataMatrix, DenseMatrix(DenseVector(dataArray)))
    i += 1 // без счетчика у меня напроч умирал юпитер, видимо потому, что пытался вывести все значения матрицы 
}



dataFile: io.BufferedSource = empty iterator
dataMatrix: DenseMatrix[Double] = 634995.0  0.0  463.0  1.0  0.0   0.0   0.0   0.0   0.0    65.0  ... (28 total)
634995.0  0.0  463.0  1.0  0.0   0.0   0.0   0.0   0.0    10.0  ...
634995.0  0.0  463.0  1.0  0.0   0.0   0.0   0.0   0.0    14.0  ...
634995.0  0.0  463.0  1.0  7.0   0.0   3.0   7.0   -3.0   62.0  ...
634995.0  0.0  463.0  1.0  1.0   0.0   0.0   1.0   0.0    58.0  ...
634995.0  0.0  463.0  1.0  0.0   0.0   0.0   0.0   0.0    60.0  ...
634995.0  0.0  463.0  1.0  0.0   0.0   0.0   0.0   0.0    68.0  ...
634995.0  0.0  463.0  1.0  1.0   0.0   1.0   1.0   -1.0   32.0  ...
634995.0  0.0  463.0  1.0  0.0   0.0   0.0   0.0   0.0    35.0  ...
634995.0  0.0  463.0  1.0  0.0   0.0   0.0   0.0   0.0    48.0  ...
634995.0  0.0  463.0  1.0  0.0   0.0   0.0   0.0   0.0    52.0  ...
634995.0  0.0  463.0  1.0  1.0   0.0   0.0   1.0   0.0    69.0  ...
634995.0  0.0  463.0  1.0  0.0   0.0   0.0   0.0   0.0    3.0   ...
634995.0  0.0  463.0  1.0  1.0   1.0   0.0   1.0   1.0    37.0  ...
634995.0  0.0  463.0  1.0  0.0   0.0   0.0   0.0   0.0    23.0  ...
634995.0  0.0  463.0  1.0  3.0   0.0   3.0   3.0   -3.0   40.0  ...
634995.0  0.0  463.0  1.0  3.0   0.0   3.0   2.0   -3.0   54.0  ...
634995.0  0.0  463.0  1.0  0.0   0.0   0.0   0.0   0.0    29.0  ...
634995.0  0.0  463.0  1.0  1.0   0.0   1.0   1.0   -1.0   36.0  ...
634995.0  0.0  463.0  1.0  0.0   0.0   0.0   0.0   0.0    11.0  ...
634995.0  0.0  463.0  1.0  0.0   0.0   0.0   0.0   0.0    62.0  ...
634995.0  0.0  463.0  1.0  77.0  46.0  31.0  74.0  15.0   31.0  ...
634995.0  0.0  463.0  1.0  6.0   0.0   1.0   6.0   -1.0   54.0  ...
634995.0  0.0  463.0  1.0  0.0   0.0   0.0   0.0   0.0    69.0  ...
634995.0  0.0  463.0  1.0  0.0   0.0   0.0   0.0   0.0    71.0  ...
634995.0  0.0  463.0  1.0  0.0   0.0   0.0   0.0   0.0    61.0  ...
634995.0  0.0  463.0  1.0  4.0   4.0   0.0   4.0   4.0    13.0  ...
634995.0  0.0  463.0  1.0  29.0  3.0   11.0  25.0  -8.0   49.0  ...
634995.0  0.0  463.0  1.0  3.0   1.0   2.0   3.0   -1.0   35.0  ...
634995.0  0.0  463.0  1.0  2.0   2.0   0.0   2.0   2.0    11.0  ...
634995.0  0.0  463.0  1.0  2.0   0.0   2.0   2.0   -2.0   37.0  ...
634995.0  0.0  463.0  1.0  0.0   0.0   0.0   0.0   0.0    0.0   ...
634995.0  0.0  463.0  1.0  2.0   2.0   0.0   2.0   2.0    11.0  ...
634995.0  0.0  463.0  1.0  0.0   0.0   0.0   0.0   0.0    43.0  ...
634995.0  0.0  463.0  1.0  2.0   2.0   0.0   2.0   2.0    18.0  ...
634995.0  0.0  463.0  1.0  0.0   0.0   0.0   0.0   0.0    64.0  ...
634995.0  0.0  463.0  1.0  0.0   0.0   0.0   0.0   0.0    51.0  ...
634995.0  0.0  463.0  1.0  0.0   0.0   0.0   0.0   0.0    67.0  ...
634995.0  0.0  463.0  1.0  8.0   8.0   0.0   8.0   8.0    12.0  ...
...
i: Int = 40949

In [3]:
// разбиваем данные
val X_all = dataMatrix(0 to dataMatrix.rows - 1, 0 to dataMatrix.cols-2)
val Y_all = dataMatrix(::,27)

X_all: DenseMatrix[Double] = 634995.0  0.0  463.0  1.0  0.0   0.0   0.0   0.0   0.0    65.0  ... (27 total)
634995.0  0.0  463.0  1.0  0.0   0.0   0.0   0.0   0.0    10.0  ...
634995.0  0.0  463.0  1.0  0.0   0.0   0.0   0.0   0.0    14.0  ...
634995.0  0.0  463.0  1.0  7.0   0.0   3.0   7.0   -3.0   62.0  ...
634995.0  0.0  463.0  1.0  1.0   0.0   0.0   1.0   0.0    58.0  ...
634995.0  0.0  463.0  1.0  0.0   0.0   0.0   0.0   0.0    60.0  ...
634995.0  0.0  463.0  1.0  0.0   0.0   0.0   0.0   0.0    68.0  ...
634995.0  0.0  463.0  1.0  1.0   0.0   1.0   1.0   -1.0   32.0  ...
634995.0  0.0  463.0  1.0  0.0   0.0   0.0   0.0   0.0    35.0  ...
634995.0  0.0  463.0  1.0  0.0   0.0   0.0   0.0   0.0    48.0  ...
634995.0  0.0  463.0  1.0  0.0   0.0   0.0   0.0   0.0    52.0  ...
634995.0  0.0  463.0  1.0  1.0   0.0   0.0   1.0   0.0    69.0  ...
634995.0  0.0  463.0  1.0  0.0   0.0   0.0   0.0   0.0    3.0   ...
634995.0  0.0  463.0  1.0  1.0   1.0   0.0   1.0   1.0    37.0  ...
634995.0

In [4]:
// Посмотрим минимальное, максимальное и среднее значения целевого признака

import breeze.stats.mean
println("Min value = " + Y_all.min)
println("Max value = " + Y_all.max)
println("Mean value = " + mean(Y_all))


Min value = 0.0
Max value = 1305.0
Mean value = 7.322889447849774


import breeze.stats.mean


### Посмотрим на корреляцию в данных (не осилил это сделать в скале, сделал через LibreOffice)

##### для удобства разбил на две матрицы корреляции, чтобы нормально влезло в экран

![title](dataset_corr.png)


Как мы можем заметить, самая высока степерь корреляции с целевым признаком output у признака **comm24**. Далее по списку идут: **diff2448 , comm24_1 , commBase**. Их и будем использовать в качестве описываемых признаков.



In [5]:
// вырезаем необходимые признаки (4,5,7,8), ормализуем их и собираем новую матрцу
import breeze.optimize._

val X_commBase = DenseMatrix(X_all(::,4) / norm(X_all(::,4)))
val X_comm24 = DenseMatrix(X_all(::,5) / norm(X_all(::,5)))
val X_comm24_1 = DenseMatrix(X_all(::,7) / norm(X_all(::,7)))
val X_diff2448 = DenseMatrix(X_all(::,8) / norm(X_all(::,8)))
var X = DenseMatrix.zeros[Double](X_commBase(::,0).length,0)

X = DenseMatrix.horzcat(X_commBase.t, X_comm24.t)
X = DenseMatrix.horzcat(X, X_comm24_1.t)
X = DenseMatrix.horzcat(X, X_diff2448.t)

import breeze.optimize._


X_commBase: DenseMatrix[Double] = 0.0  0.0  0.0  2.3392997780626447E-4  3.3418568258037784E-5  ... (40949 total)
X_comm24: DenseMatrix[Double] = 0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  ... (40949 total)
X_comm24_1: DenseMatrix[Double] = 0.0  0.0  0.0  2.496481321768704E-4  3.5664018882410056E-5  ... (40949 total)
X_diff2448: DenseMatrix[Double] = 0.0  0.0  0.0  -1.573196091867818E-4  0.0  0.0  0.0  ... (40949 total)
X: DenseMatrix[Double] = 0.0                    0.0                    0.0                    0.0                     
0.0                    0.0                    0.0                    0.0                     
0.0                    0.0                    0.0                    0.0                     
2.3392997780626447E-4  0.0                    2.496481321768704E-4   -1.573196091867818E-4   
3.3418568258037784E-5  0.0                    3.5664018882410056E-5  0.0                     
0.0                    0.0            

In [6]:
// разделим выборки на train и test
val test_count = (X.rows * 0.67).toInt
val X_train = X(0 to test_count, 0 to X.cols - 1)
val X_test = X(test_count to X.rows - 1, 0 to X.cols - 1)
val Y_train = Y_all(0 to test_count)
val Y_test = Y_all(test_count to Y_all.length - 1)
assert(X_train.rows == Y_train.length)
assert(X_test.rows == Y_test.length)

test_count: Int = 27435
X_train: DenseMatrix[Double] = 0.0                    0.0                    0.0                    0.0                     
0.0                    0.0                    0.0                    0.0                     
0.0                    0.0                    0.0                    0.0                     
2.3392997780626447E-4  0.0                    2.496481321768704E-4   -1.573196091867818E-4   
3.3418568258037784E-5  0.0                    3.5664018882410056E-5  0.0                     
0.0                    0.0                    0.0                    0.0                     
0.0                    0.0                    0.0                    0.0                     
3.3418568258037784E-5  0.0                    3.5664018882410056E-5  -5.243986972892727E-5   
0.0                    0.0                    0.0                    0.0                     
0.0                    0.0                    0.0                    0.0                     
0.0  

In [7]:
// применяем регрессионные модели на 4х колонках
import breeze.stats.regression.{leastSquares, lasso}
val squaresModel = leastSquares(X_train, Y_train)
val lassoModel = lasso(X_train, Y_train, 0.01)


Apr 12, 2021 6:07:11 PM com.github.fommil.netlib.LAPACK <clinit>
Apr 12, 2021 6:07:11 PM com.github.fommil.netlib.LAPACK <clinit>


import breeze.stats.regression.{leastSquares, lasso}

squaresModel: breeze.stats.regression.LeastSquaresRegressionResult = LeastSquaresRegressionResult(
  DenseVector(4118.741085583476, 4451.497358203682, -4288.017965150031, 323.5540881731232),
  2.515324777289502E7
)
lassoModel: breeze.stats.regression.LassoResult = LassoResult(
  DenseVector(3185.500520476411, 4363.898662961942, -3292.4296418889417, 344.39329994776625),
  2.515581780504165E7,
  0.01
)

In [8]:
// применим наши модели на тестовой выборке

assert(X_test.cols == squaresModel.coefficients.length)
val Y_pred_squaresModel = X_test * squaresModel.coefficients
val Y_pred_lassoModel = X_test * lassoModel.coefficients


Apr 12, 2021 6:07:12 PM com.github.fommil.netlib.BLAS <clinit>
Apr 12, 2021 6:07:12 PM com.github.fommil.netlib.BLAS <clinit>


Y_pred_squaresModel: DenseVector[Double] = DenseVector(-4.023403509631854, 0.8388782378768742, 9.65288517157837, -6.497753453038508, 74.46511418074743, 5.196706061385702, -6.649537569554759, 277.1757027838933, 0.0, 0.6165695020861552, -2.786635348816174, -0.56819541996922, -0.32756365772378615, -1.1763781537086588, 1.0803743403395907, 1.2532712203995722, 2.1775668731470796, -0.3838196999492984, 11.307665486706096, 21.272644292420278, 0.6060988006463117, -0.05066215249294603, 28.131265357171255, -2.3409812215473593, 0.011995020208825709, 26.752281761231494, 0.34842413534243083, 19.029973623968797, 33.837298676979266, 11.583462205894051, 2.0183656344381475, 18.478380185592886, 3.5853573494433952, -0.509446013524619, 51.5739864881473, -3.3415380346997594, -0.29531099991825727, -0.6826533737225802, 38.61154068631348, 27.209131303538605, -0.5127442692065393, 0.018875466534922003, -3.7420284053498887, 14.89302283614949, 5.240137664571117, 13.103365277449486, 0.038849169169367315, 68.39758635

### В качестве функции потерь будем использовать MAE т.к. нам необходимо понимать, на сколько комментариев мы промахиваемся в ту или иную сторону

In [9]:
import math.abs
val mae = (X: DenseVector[Double], X_pred: DenseVector[Double]) => {
    ((X - X_pred).map(x => abs(x)).sum) / X.length
}

val mae_test_squaresModel = mae(Y_test, Y_pred_squaresModel)
val mae_test_lassoModel = mae(Y_test, Y_pred_lassoModel)


println(mae_test_squaresModel)
println(mae_test_lassoModel)

6.231902929890511
6.188256420491942


import math.abs

mae: (DenseVector[Double], DenseVector[Double]) => Double = ammonite.$sess.cmd8$Helper$$Lambda$2716/702463059@2b981b7
mae_test_squaresModel: Double = 6.231902929890511
mae_test_lassoModel: Double = 6.188256420491942

##### Судя по результатам оценки, lassoModel показывает немного большую точность чем squaresModel на текущих данных

## Вычислим коэфиценты своими силами и сравним результаты

### Т.к. данных у нас не так много, можно использовать самую простую реализацию

In [10]:
import breeze.linalg.inv
val X_train_t = X_train.t
val coef_custom = inv((X_train_t * X_train)) * X_train_t * Y_train
var Y_pred_custom = X_test * coef_custom
val mae_test_custom = mae(Y_test, Y_pred_custom)

println(mae_test_custom)

6.231902929896886


import breeze.linalg.inv

X_train_t: DenseMatrix[Double] = 0.0  0.0  0.0  2.3392997780626447E-4  3.3418568258037784E-5  ... (27436 total)
0.0  0.0  0.0  0.0                    0.0                    ...
0.0  0.0  0.0  2.496481321768704E-4   3.5664018882410056E-5  ...
0.0  0.0  0.0  -1.573196091867818E-4  0.0                    ...
coef_custom: DenseVector[Double] = DenseVector(4118.741085726199, 4451.497358212116, -4288.017965298979, 323.55408817308734)
Y_pred_custom: DenseVector[Double] = DenseVector(-4.023403509713147, 0.8388782378883921, 9.652885171577497, -6.497753453186885, 74.46511418074068, 5.1967060613705565, -6.649537569786064, 277.17570278386813, 0.0, 0.6165695020730544, -2.786635348902501, -0.5681954200120346, -0.3275636577358231, -1.176378153790047, 1.0803743403372092, 1.253271220388045, 2.1775668731461675, -0.38381969996231585, 11.307665486705071, 21.272644292328554, 0.6060988006148775, -0.0506621525041403, 28.131265357168704, -2.34098122161337, 0.011995020210344258, 26.75

### Итоговые значения средних абсолютных ошибок: 

* lassoModel = 6.188256420491942
* squaresModel = 6.231902929890511
* custom = 6.231902929896886

Кастомная реализация показала практически тот же результат, что и коробочные модели. Все модели имеют погрешность прмиерно равную 6.2 

Учитывая, что среднее значение целевой переменной = 7.32 , результат мягко говоря не радует. Верояно были выбраны неподходящие признаки т.к. опираться на одну лишь корреляцию не совсем верно. Нужен более глубокий анализ, да и описание данных тоже не помешало бы т.к. не имея описания каждой колонки не совсем ясен смысл данных.